### PR TITLE
IGNITE-25572 Configuration compatibility. Support Polymorphic configuration snapshot

### PR DESCRIPTION
**Pull request by lowka**
---

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/runner/src/test/java/org/apache/ignite/internal/configuration/compatibility/framework/ConfigurationTreeComparator.java [hunk 1]
```diff
@@ -227,15 +254,73 @@ private static boolean rootsMatch(ConfigNode candidate, ConfigNode current) {
...
         if (!context.errors.isEmpty()) {
             return;
         }
-        compareNodes(candidate, current, context);
+
+        if (candidate.isPolymorphic() && current.isPolymorphic()) {
+            if (!Objects.equals(candidate.defaultPolymorphicInstanceId(), current.defaultPolymorphicInstanceId())) {
+                ConfigNode anyNode = candidate.nodes().values().iterator().next();
+                String message = format(
+                        "Default polymorphic instance id does not match. Expected: {} got {}",
```
Comment (ibessonov) [inline modules/runner/src/test/java/org/apache/ignite/internal/configuration/compatibility/framework/ConfigurationTreeComparator.java]: Is this a mistake? I don't think so, we may want to switch the default implementation of a failure handler, and it's supposed to be a valid change

**Review by ibessonov — APPROVED** [review]

File: modules/runner/src/test/java/org/apache/ignite/internal/configuration/compatibility/framework/ConfigurationTreeScanner.java [hunk 2]
```diff
@@ -83,33 +81,75 @@ public class ConfigurationTreeScanner {
...
+
+            // Collect nodes that correspond to polymorphic instances
+            for (Class<?> instanceClass : instanceClasses) {
+                ConfigNode instanceTypeNode = createNodeForField(currentNode, field, instanceClass);
+                polymorphicInstances.put(instanceTypeNode.polymorphicInstanceId(), instanceTypeNode);
+
+                // Each polymorphic instance includes fields from the base class
+                scan(instanceTypeNode, field.getType(), context);
+                // And its own fields ignoring base class fields.
+                scan(instanceTypeNode, instanceClass, context, baseClassFields);
```
Comment (AMashenkov) [inline modules/runner/src/test/java/org/apache/ignite/internal/configuration/compatibility/framework/ConfigurationTreeScanner.java:148]: Should we scan all PolymorphicInstance class parents instead?
E.g. there can be intermediate classes in hierarchy between `field.getType()` and `instanceClass`.

Comment (AMashenkov) [reply modules/runner/src/test/java/org/apache/ignite/internal/configuration/compatibility/framework/ConfigurationTreeScanner.java:148]: @ibessonov WDYT?

Comment (ibessonov) [reply modules/runner/src/test/java/org/apache/ignite/internal/configuration/compatibility/framework/ConfigurationTreeScanner.java:148]: I don't think it's allowed right now, but relying on a hierarchy of Java classes instead of a hierarchy of "logical" configuration schema representation might not be a good idea, IF I understood your point correctly

**Review by AMashenkov — APPROVED** [review]
